lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+            parser::MessageFormattedText{
+                "ELSEWHERE statement name mismatch"_err_en_US});
+      }
+    }
+  }
+
+  /// \brief Check constraint <i>construct-name</i> in scope (C1134 and C1166)
+  /// \param SStr  a string to specify the statement, \c CYCLE or \c EXIT
+  /// \param Label the name used by the \c CYCLE or \c EXIT
+  template<typename A>
+  void CheckLabelContext(const char *const stmtString, const A &constructName) {
+    const auto I{std::find(
+        constructNames_.crbegin(), constructNames_.crend(), constructName)};
+    if (I == constructNames_.crend()) {
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "%s construct-name '%s' is not in scope"_err_en_US, stmtString,
+              constructName.c_str()});
+    }
+  }
+
+  /// \brief Check label range
+  /// Constraint per section 6.2.5, paragraph 2
+  void CheckLabelInRange(parser::Label label) {
+    if ((label < 1) || (label > 99999)) {
+      // this is an error: labels must have a value 1 to 99999, inclusive
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "label '%lu' is out of range"_err_en_US, label});
+    }
+  }
+
+  /// \brief Add a labeled statement (label must be distinct)
+  /// Constraint per section 6.2.5., paragraph 2
+  void AddTrgt(parser::Label label, TargetStmtType targetStmtType) {
+    CheckLabelInRange(label);
+    const auto pair{programUnits_.back().targetStmts_.insert(
+        {label, {currentScope_, currentPosition_, targetStmtType}})};
+    if (!pair.second) {
+      // this is an error: labels must be pairwise distinct
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "label '%lu' is not distinct"_err_en_US, label});
+    }
+    // Don't enforce a limit to the cardinality of labels
+  }
+
+  /// \brief Reference to a labeled statement from a DO statement
+  void AddDoBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().doStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief Reference to a labeled FORMAT statement
+  void AddFmtBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().formatStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief Reference to a labeled statement as a (possible) branch
+  void AddBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().otherStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief References to labeled statements as (possible) branches
+  void AddBase(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      AddBase(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;  ///< results for each program unit
+  ErrorHandler eh;  ///< error handler, collects messages
```

Comment:
`eh` should be `eh_`.

Reply:
The trailing underscore serves no purpose in my opinion.

Reply:
That's just the convention for private data members.

Reply:
added